### ISSUE TITLE
Rectangle roi not working with non-square pixel

### ISSUE THREAD

**Issue by ericpre** [hyperspy/hyperspy#1693]
The following is not working:

```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.arange(2 * 4 * 6).reshape(2, 4, 6))
s.axes_manager[0].scale = 0.5
#s.axes_manager[1].scale = 0.5

roi_nav = hs.roi.RectangularROI(0, 0, 1, 0.5)`
s.plot()
w = roi_nav.add_widget(s)
```
and produces the following error:

```python
Traceback (most recent call last):

  File "<ipython-input-1-83892adf2fe2>", line 12, in <module>
    w = roi_nav.add_widget(s)

  File "/home/eric/Python_dev/hyperspy/hyperspy/roi.py", line 447, in add_widget
    widget.set_mpl_ax(ax)

  File "/home/eric/Python_dev/hyperspy/hyperspy/drawing/widget.py", line 168, in set_mpl_ax
    self.select()

  File "/home/eric/Python_dev/hyperspy/hyperspy/drawing/widget.py", line 182, in select
    canvas.pick_event(mouseevent, self.patch[0])

  File "/opt/anaconda3/lib/python3.6/site-packages/matplotlib/backend_bases.py", line 1883, in pick_event
    self.callbacks.process(s, event)

  File "/opt/anaconda3/lib/python3.6/site-packages/matplotlib/cbook.py", line 554, in process
    proxy(*args, **kwargs)

  File "/opt/anaconda3/lib/python3.6/site-packages/matplotlib/cbook.py", line 416, in __call__
    return mtd(*args, **kwargs)

  File "/home/eric/Python_dev/hyperspy/hyperspy/drawing/widget.py", line 671, in onpick
    super(ResizableDraggableWidgetBase, self).onpick(event)

  File "/home/eric/Python_dev/hyperspy/hyperspy/drawing/widget.py", line 438, in onpick
    super(DraggableWidgetBase, self).onpick(event)

  File "/home/eric/Python_dev/hyperspy/hyperspy/drawing/widget.py", line 958, in onpick
    self.pick_offset = (x - self._pos[0], y - self._pos[1])

TypeError: unsupported operand type(s) for -: 'NoneType' and 'float'
```

If you uncomment the line:

```python
s.axes_manager[1].scale = 0.5
```
Then, this is working.

**Comment by ericpre** [comment]
Slightly different issue:

```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.arange(2 * 4 * 6).reshape(2, 4, 6))
s.axes_manager[0].scale = 0.5
s.axes_manager[1].scale = 0.5

roi_nav = hs.roi.RectangularROI(0, 0, 2, 0.5)

s.plot()

w = roi_nav.add_widget(s)
```
makes the error:

```python
Traceback (most recent call last):

  File "<ipython-input-3-04b533228dbb>", line 12, in <module>
    w = roi_nav.add_widget(s)

  File "/home/eric/Python_dev/hyperspy/hyperspy/roi.py", line 438, in add_widget
    self._apply_roi2widget(widget)

  File "/home/eric/Python_dev/hyperspy/hyperspy/roi.py", line 764, in _apply_roi2widget
    right=self.right, top=self.top)

  File "/home/eric/Python_dev/hyperspy/hyperspy/drawing/_widgets/rectangles.py", line 151, in set_bounds
    raise ValueError('`top` value is not in range.')

ValueError: `top` value is not in range.
```
The error message is a change of my local branch; in the upstream branches, the message is empty.

**Comment by francisco-dlp** [comment]
I think that the ROI is working with non-square pixels. The issue with your first example is that you're trying to create a half-pixel height ROI. Nevertheless, I can't reproduce the error, but the ROI that I gets approximated to a 0-height one when I try to move it which may be considered a bug.

The second issue only appears when adding the widget. When adding a thinner widget and resizing it to width 2 it behaves as expected, so I guess that there is an issue with the validation only when adding it.

Do you agree with my comments? If yes or partially yes, could you update the title of the issue?

**Comment by ericpre** [comment]
I was a bit confused because they may be actually several issue convolved here... my frustration pushed me to through the error without much comments!! ;)

This is surprising and annoying that you can't reproduce the first issue. In my case, the half-pixel roi does not seem to raise an issue. Below, I corrected the half-pixel roi and I still get the same issue, while for larger dataset, it works:
```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.arange(2 * 4 * 6).reshape(2, 4, 6))
s.axes_manager[0].scale = 0.5
#s.axes_manager[1].scale = 0.5

roi_nav = hs.roi.RectangularROI(0, 0, 0.5, 1)
s.plot()
w = roi_nav.add_widget(s)
```
Same error as the first issue reported here. While the following works nicely:

```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.arange(4 * 8 * 6).reshape(4, 8, 6))
s.axes_manager[0].scale = 0.5
#s.axes_manager[1].scale = 0.5

roi_nav = hs.roi.RectangularROI(0, 0, 0.5, 1)
s.plot()
w = roi_nav.add_widget(s)
```
I put this for the record, but if you can't reproduce, I may it (for now at least). I will change the title of the issue, once we get a better idea of what is exactly the issue.

**Comment by thomasaarholt** [comment]
@ericpre Neither of those two blocks of code give me an error.

**Comment by ericpre** [comment]
Thanks @thomasaarholt for looking at this. 

Travis and appveyor can both reproduce, see some of the tests added in #1674 (commit 7bf9df2) and the corresponding [Travis](https://travis-ci.org/hyperspy/hyperspy/builds/256105316) and [appveyor](https://ci.appveyor.com/project/hyperspy/hyperspy/build/1.0.2539) logs.

**Comment by francisco-dlp** [comment]
Fixed in #1699
